espnet2/bin/diar_inference.py

```diff
@@ -0,0 +1,239 @@
+                    waves[:, :, -overlap_length:],
+                    enh_waves[i][:, :, :overlap_length],
+                    criterion="si_snr",
+                )
+                # repermute separated streams in current segment
+                for batch in range(batch_size):
+                    enh_waves[i][:, batch] = enh_waves[i][perm[batch], batch]
+
+                if i == num_segments - 1:
+                    enh_waves[i][:, :, t:] = 0
+                    enh_waves_res_i = enh_waves[i][:, :, overlap_length:t]
+                else:
+                    enh_waves_res_i = enh_waves[i][:, :, overlap_length:]
+
+                # overlap-and-add (average over the overlapped part)
+                waves[:, :, -overlap_length:] = (
+                    waves[:, :, -overlap_length:] + enh_waves[i][:, :, :overlap_length]
+                ) / 2
+                # concatenate the residual parts of the later segment
+                waves = torch.cat([waves, enh_waves_res_i], dim=2)
+            # ensure the stitched length is same as input
+            assert waves.size(2) == speech_mix.size(1), (waves.shape, speech_mix.shape)
+            waves = torch.unbind(waves, dim=0)
+        else:
+            # b. Enhancement/Separation Forward
+            feats, f_lens = self.enh_model.encoder(speech_mix, lengths)
+            feats, _, _ = self.enh_model.separator(feats, f_lens)
+            waves = [self.enh_model.decoder(f, lengths)[0] for f in feats]
+
+        assert len(waves) == self.num_spk, len(waves) == self.num_spk
+        assert len(waves[0]) == batch_size, (len(waves[0]), batch_size)
+        if self.normalize_output_wav:
+            waves = [
+                (w / abs(w).max(dim=1, keepdim=True)[0] * 0.9).cpu().numpy()
+                for w in waves
+            ]  # list[(batch, sample)]
+        else:
+            waves = [w.cpu().numpy() for w in waves]
+
+        return waves
+
+    @torch.no_grad()
+    def cal_permumation(self, ref_wavs, enh_wavs, criterion="si_snr"):
+        """Calculate the permutation between seaprated streams in two adjacent segments.
+
+        Args:
+            ref_wavs (List[torch.Tensor]): [(Batch, Nsamples)]
+            enh_wavs (List[torch.Tensor]): [(Batch, Nsamples)]
+            criterion (str): one of ("si_snr", "mse", "corr)
+        Returns:
+            perm (torch.Tensor): permutation for enh_wavs (Batch, num_spk)
+        """
+        loss_func = {
+            "si_snr": self.enh_model.si_snr_loss,
+            "mse": lambda enh, ref: torch.mean((enh - ref).pow(2), dim=1),
+            "corr": lambda enh, ref: (
+                (enh * ref).sum(dim=1)
+                / (enh.pow(2).sum(dim=1) * ref.pow(2).sum(dim=1) + EPS)
+            ).clamp(min=EPS, max=1 - EPS),
+        }[criterion]
+
+        _, perm = self.enh_model._permutation_loss(ref_wavs, enh_wavs, loss_func)
+        return perm
+
+
+def humanfriendly_or_none(value: str):
+    if value in ("none", "None", "NONE"):
```

Comment:
Could you use it?

https://github.com/espnet/espnet/blob/2cfecaf70a2d2826e43db9a2aa34305c6628c5a5/espnet2/utils/types.py#L75